Several shader preprocessor parser fixes and improvements

bitsawer opened the pull request:
Fixes several small issues with the shader preprocessor.

* Add support for line continuation ("\\" + "\n"). The official docs (https://docs.godotengine.org/en/latest/tutorials/shaders/shader_reference/shader_preprocessor.html) show a code snippet where it's supported, but it actually doesn't work because I forgot to code it. With this PR it's now supported in most places and this snippet now works:

``` cpp
// Multiline replacement using backslashes for continuation:
#define SAMPLE(param1, param2, param3, param4) long_function_call( \
        param1, \
        param2, \
        param3, \
        param4 \
)

```

* Parser will now raise an error if there are extra non-whitespace characters after #else, #endif or #undef XYZ. Added consume_empty_line() for this to make the intent clear. Previously this was valid:

``` cpp
#endif some junk that is ignored
#undef SOMETHING gga8ha5e978 hs5sv5
```

* Fix an #include throwing off error line numbers by one in the editor end error messages. Also fixes some other line number offset errors.
* #undef no longer causes an error when used on a non-existing macro. This is similar to other preprocessors.

Technically you could say that this PR breaks compatibility, but in practice I don't think anyone will notice because anything that breaks was wrong and didn't do anything to begin with.

bitsawer committed:
Several shader preprocessor parser fixes and improvements

clayjohn commented:
The changes seem fine to me, but I can't really comment on the code itself as I am pretty unfamiliar with the preprocessor.

Chaosus commented:
Looks great. For some reason, the following code does not parse properly:

```
shader_type spatial;

vec3 long_function_call(vec3 param1)
{ 
 param1 += vec3(0, 0, 1); 
 return param1; 
}

#define SAMPLE(param1) long_function_call( \
    param1 \
)

void fragment() {
	ALBEDO = SAMPLE(vec3(1.0, 0.0, 0.0));
}
```
![image](https://user-images.githubusercontent.com/3036176/214770702-a0b44b15-ff27-46ae-99ce-f585504b4974.png)

(It should be correct according to the ShaderToy)

bitsawer commented:
@Chaosus 

Looks like that macro argument expansion never supported very complex arguments (with commas or parentheses) to begin with, I guess you are just the first one to notice it and report the issue. I fixed that now and updated the PR.

Previously it simply used String.split(",") to split the arguments, which caused it to see three arguments as `SAMPLE(vec3(1.0, 1.0, 0.0))` turned into

* vec3(1.0
* 1.0
* 0.0

instead of the correct single argument `vec3(1.0, 1.0, 0.0)`. I replaced the code with better parsing, should work properly now.

Chaosus reviewed (approved):
LGTM

akien-mga commented:
Thanks!
